tools/firmware.toit

```diff
@@ -1357,13 +1350,97 @@ class Envelope:
       throw "cannot open envelope - expected version $ENVELOPE-FORMAT-VERSION, was $version"
     return version
 
+class RelocationInformation:
+  relocation-bytes/ByteArray
+  word-size/int  // In bytes.
+
+  constructor.from relocatable/ByteArray --.word-size:
+    chunk-size := get-relocatable-chunk-byte-size --word-size=word-size
+    chunk-count := ceil_ relocatable.size chunk-size
+    relocation-bytes = ByteArray chunk-count * word-size
+    for i := 0; i < chunk-count; i++:
+      relocation-bytes.replace (i * word-size) relocatable[i * chunk-size..i * chunk-size + word-size]
+
+  make-relocatable image/ByteArray -> ByteArray:
+    relocatable-chunk-size := get-relocatable-chunk-byte-size --word-size=word-size
+    relocated-chunk-size := get-relocated-chunk-byte-size --word-size=word-size
+    result := ByteArray (ceil_ image.size relocated-chunk-size) * relocatable-chunk-size
+
+    image-i := 0
+    result-i := 0
+    relocation-i := 0
+    while image-i < image.size:
+      relocation-info := relocation-i < relocation-bytes.size
+          ? relocation-bytes[relocation-i..relocation-i + word-size]
+          : ByteArray word-size  // We want to be able to attach assets to the image.
+      relocation-i += word-size
+
+      chunk := image[image-i..min image.size (image-i + relocated-chunk-size)]
+      image-i += chunk.size
+
+      result.replace result-i relocation-info
+      result-i += relocation-info.size
+      result.replace result-i chunk
+      result-i += chunk.size
+
+    return result
+
+  static relocated-size relocatable-size/int --word-size/int -> int:
+    relocatable-chunk-size := get-relocatable-chunk-byte-size --word-size=word-size
+    relocated-chunk-size := get-relocated-chunk-byte-size --word-size=word-size
+    chunk-count := ceil_ relocatable-size relocatable-chunk-size
+    // One word of every chunk is used for relocation information.
+    return chunk-count * relocated-chunk-size
+
+  static get-relocatable-chunk-byte-size --word-size/int -> int:
+    word-bit-size := word-size * 8
+    // One word for the relocation-information, followed by one word for each bit of it.
+    chunk-word-size := 1 + word-bit-size
+    return chunk-word-size * word-size
+
+  static get-relocated-chunk-byte-size --word-size/int -> int:
+    word-bit-size := word-size * 8
+    return word-size * word-bit-size
+
+ceil_ x/int y/int -> int:
+  return (x + y - 1) / y
+
 class ContainerEntry:
   id/uuid.Uuid
   name/string
   flags/int
   relocatable/ByteArray
   assets/ByteArray?
-  constructor .id .name .relocatable --.flags --.assets:
+  word-size/int
+
+  constructor .id .name .relocatable --.flags --.assets --.word-size:
+
+  relocated-size -> int:
+    return relocatable.size - (RelocationInformation.relocated-size relocatable.size --word-size=word-size)
+
+  cook --relocation-base/int --attach-assets/bool --system-uuid/uuid.Uuid -> ByteArray:
```

Comment:
Maybe this should just be called `relocate`? Or `make-relocated` to fit with `make-relocatable`?